text/1164034/phase_king.tex

```diff
@@ -0,0 +1,78 @@
+\begin{itemize}
+    \item \textbf{Agreement} All correct processes must agree on the same (single) value.
+    \item \textbf{Validity} If all the correct processes have the same initial value, then the agreed upon value by all the correct processes must be that same value.
+    \item \textbf{Termination} Each correct process must eventually decide on a value. 
+\end{itemize}
+\end{defn}
+
+One can measure the quality of protocols solving the consensus problem using the following parameters: total number of processors $n$, number of rounds of message exchange $r$ and maximal message length $m$. Their optima are known to be respectively $3t+1, t+1$ and $1$.
+
+The King Phase protocol, proposed by Berman, Garay and Perry, is one of protocols solving the consensus problem. It's also asymptotically optimal in all the three parameters mentioned earlier. Namely, their values for the protocol are respectively $3t+1$, $3t+3$ and $2$.
+
+\subsection*{Protocol}
+The idea of the protocol is as follows. It runs $t+1$ phases, each consisting of three exchange rounds. Each processor has a local variable $V$ which can obtain values from $\{0,1,2\}$ and two arrays $C$ and $D$ counting messages with different values received from the other processors in the first and the second exchange round respectively. Variables $V$ are set to the initial values at the beginning. For each phase, if all the correct processors are in agreement (have the same values $V$) before given phase, then the agreement will be preserved (all the processors will have the same $V$ values) after this phase. This fact not only assures the validity, but also implies it's enough to reach the agreement once, in any phase. With that knowledge let's analyse a phase. The first round set all the correct processors' $V$ variables to merely zeroes and twos or to merely ones and twos, beacuse if there was a pair of correct processors having $V=1$ and $V=2$, then $n\ge2(n-t)-t$, but $n\geq 3t+1$, which would mean $3t\ge n \ge 3t+1$, which is a contradiction. The other two rounds allow to reach consensus between all the correct processors. In each phase a unique processor become a king of the phase (the phase number determines the id of the king). If the king is correct, then he broadcast his $V$. Two cases may occur now. Either all the correct processors set $V$ to the same value from $\{0,1\}$ in the third round or there is some for which $V\neq2$ and $D(V)\ge n-t$. The later case implies that for each of the correct processor having $V\neq2$ follows $D(V)>t$ (because there are $t$ faulty processors and $n\geq 3t+1$), meaning that the correct processors have $V$ already set to the same value from $\{0,1\}$ after the second phase. At least one of $t+1$ kings is correct, so combining our observations we derive the valid protocol for the consensus problem.
```

Comment:
Please extract to the separate lemmas the following claims and provide proof sketches for better readability (and for sure it'll become handy in the future):
1. if all correct nodes have the same values before the phase, then it holds also after the phase,
2. if a king is genuine, then after its round all correct nodes have the same values,
3. if the first correct node terminates in round $k$, then all other correct nodes terminate at most in round $k + 1$.